Fix breaking avatar on change of email id

adnrs96 opened the pull request:
This fixes issue #3776.
This introduces changes to make avatar storage based on user id's instead of email id's.

smarx commented:
*Automated message from Dropbox CLA bot*

@adnrs96, it looks like you've already signed the Dropbox CLA. Thanks!

timabbott commented on zerver/migrations/0055_move_avatars_to_be_uid_based.py:
```diff
@@ -0,0 +1,34 @@
+        if avatar_connection.status_code == 200:
+            avatar_data = avatar_connection.content
+            avatar_file = SimpleUploadedFile('avatar.jpg', avatar_data, 'image/jpeg')
+            upload_avatar_image(avatar_file, user_profile, user_profile.id)
```
I would prefer to do something that preserves any metadata on the image object, rather than doing a full fetch+re-upload, if it's reasonable to do that...

adnrs96 replied:
I tried to do that for S3 now.

timabbott commented on zerver/lib/upload.py:
```diff
@@ -381,18 +381,18 @@ def delete_message_image(self, path_id):
 
         resized_medium = resize_avatar(image_data, MEDIUM_AVATAR_SIZE)
-        write_local_file('avatars', email_hash+'-medium.png', resized_medium)
+        write_local_file('avatars', uid_hash+'-medium.png', resized_medium)
```
stylistically we generally do a ` ` around `+` operators; might as well fix that while we're changing this.

adnrs96 replied:
yeah will take care of this in future

adnrs96 commented on zerver/migrations/0032_verify_all_medium_avatar_images.py:
```diff
@@ -12,7 +12,7 @@ def verify_medium_avatar_image(apps, schema_editor):
     user_profile_model = apps.get_model('zerver', 'UserProfile')
     for user_profile in user_profile_model.objects.filter(avatar_source=u"U"):
-        upload_backend.ensure_medium_avatar_image(user_profile.email)
+        upload_backend.ensure_medium_avatar_image(user_profile.id)
```
yes it does makes sense. I actually didn't really thought of having email let the way it is and have ensure_medium_avatar_image call user_avatar_hash with user_profile. I mean user_avatar_path now.

adnrs96 commented:
@timabbott I have updated the PR with some new commits. I would like you to take a look at the approach now. Also the tests are not passing yet i will be investigating that currently no idea what is happening in tests.

timabbott commented:
I'm a bit confused by this commit series -- the first commit seems to start using `user_avatar_path`, but it's defined in like the 7th commit?

I should have explained this more clearly -- the purpose of doing this sort of fine-grained commit approach is that we can merge things incrementally as soon as they are correct.  But one can't do that incremental merge process if the first commit can't possibly work until the 7th commit is added -- that will break the tests for everyone!  

So the way one should be creating commits is an iterative process, where you make a change, commit it, make sure it'll pass lint/tests/etc. and brings you closer to the eventual goal, and then do the next one.  Does that make sense?

adnrs96 commented:
yeah totally, i miss understood you earlier i guess. The incremental approach should be good, let me restructure the commits.

adnrs96 commented:
@timabbott I have updated the PR with new commit's. I am afraid that i wasn't able to have divide it in more finer commits and have working stuff after every commit. Would like to have some suggestions on how that could have been done.

timabbott commented:
Awesome, I think this is very close to what I would have done!  The first commit looks great, and I just merged it.  The second commit I would actually split into 2 pieces:
* Piece 1: Does everything except that `user_avatar_path` returns the same thing that `user_avatar_hash` did.  So this is a refactor that doesn't change behavior at all, but sets up the code flow for the behavior change
* Piece 2: Changes `user_avatar_path` to return the new more complex path (and updates `test_helpers.py`); that can probably be squashed into the 3rd commit.

The advantage of that approach is that in the current 3-commit setup, commit 2 will break the app without commit 3 (because everyone's avatars will be looked for in the wrong place), whereas with the revised 3-commit version, each commit is correct.

Let me know when you've done that revision and I'll merge the rest :)

timabbott commented:
Oh, one more thought -- you should add `Fixes #1234` type lines to the end of the final commit message, so that merging this closes the various issues you fixed :).

adnrs96 commented:
Thanks for reviewing the PR and the comments. I just updated the commits with your proposed structure.

timabbott commented:
Great!  I merged the next commit as 3f58e2099396070a47d42b9b37490b23eb928ec5; the only thing I changed was fixing the import loop by moving `user_avatar_path` to `avatar_hash.py`.  We might want to rename that file, but structurally its purpose is to be the parts of the avatar library that `upload.py` imports (since `avatar.py` imports `upload.py`).

For the migration, I think you'll need to renumber it to 0060.  The other question I had is why the `LOCAL_UPLOADS_DIR` code path needs to do that download process, when I think it could just `mv` the files?

adnrs96 commented:
We may want to rename avatar_hash to avatar_extension although not much necessarily. Its just that it gives kind of suggests to have only hashing related code under it (that's why i added the import loop). 
It was super silly of me not moving the files and rather doing a download and then upload for Local Dir's.

I have updated the commit with changes in code to move the avatar's and have only avatar_hash as dependency for migration on the Zulip's code. Also i have renumbered the migration to the latest for current repo state (ie. 0060).

timabbott commented on zerver/lib/avatar_hash.py:
```diff
@@ -21,16 +21,16 @@ def gravatar_hash(email):
     return make_safe_digest(email.lower(), hashlib.md5)
 
-def user_avatar_hash(email):
+def user_avatar_hash(uid):
```
Add a comment to this function that if is modified, we need to also modify migration 0060.

timabbott commented on zerver/lib/avatar_hash.py:
```diff
@@ -21,16 +21,16 @@ def gravatar_hash(email):
-    user_email_hash = user_avatar_hash(str(user_profile.email))
-    return ('%s' % (user_email_hash))
+    user_id_hash = user_avatar_hash(str(user_profile.id))
+    return '%s/%s' % (str(user_profile.realm_id), user_id_hash)
```
Add a similar comment here.

timabbott commented on zerver/migrations/0060_move_avatars_to_be_uid_based.py:
```diff
@@ -0,0 +1,68 @@
+                            user_avatar_hash(user_profile.email))
+            bucket.delete_key(user_avatar_hash(user_profile.email) + ".original")
+            bucket.delete_key(user_avatar_hash(user_profile.email) + "-medium.png")
+            bucket.delete_key(user_avatar_hash(user_profile.email))
```
the deletes should be in a second loop so that if there's an error, it's clearer how to pick up.

timabbott commented:
I made those comment tweaks and merged this as ae31c6cc0d6a601a764c5f0a439fe6f99c9e1e8b, huge thanks for fixing this @adnrs96!

adnrs96 commented:
Thanks for doing the tweaks and merging.
